File: E-Commerce App Project (Tabbed)/Classes/Items.m
//
//  Items.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 7/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "Items.h"
#import "Item.h"

@implementation Items

+(Items *)sharedInstance{
    static Items* _sharedItems;
    
    static dispatch_once_t once;
    
    dispatch_once(&once, ^{
        _sharedItems = [[Items alloc] init];
    });
    return _sharedItems;
}


- (id)init
{
    if((self = [super init]))
    {
        _allItems = [self loadItemsFromJSON];
    }
    return self;
}

- (NSArray *)loadItemsFromJSON {
    
    NSString* filePath = [[NSBundle mainBundle] pathForResource:@"data" ofType:@"json"];
    
    NSError* error;
    NSData* jsonData =  [NSData dataWithContentsOfFile:filePath options:NSDataReadingMappedIfSafe error:&error];
    
    NSArray* jsonArray = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:nil];
    
    NSMutableArray* itemArray = [[NSMutableArray alloc] initWithCapacity:jsonArray.count];
    
    for (NSDictionary* itemDictionary in jsonArray)
    {
        Item* itemObject = [[Item alloc] init];
        
        itemObject.ID = itemDictionary[@"id"];
        itemObject.sale = itemDictionary[@"Sale"];
        itemObject.itemName = itemDictionary[@"ProductName"];
        itemObject.photoURL = itemDictionary[@"URL"];
        itemObject.price = [itemDictionary[@"Price"] doubleValue];
        itemObject.itemCategory = itemDictionary[@"ProductType"];
        itemObject.brand = itemDictionary[@"BrandName"];
        itemObject.quality = itemDictionary[@"Quality"];
        itemObject.cartAddedQuantity = 0;
        [itemArray addObject:itemObject];
    }
    return itemArray;
}
@end

File: E-Commerce App Project (Tabbed)/Classes/Items.swift
//
//  Items.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import Foundation

@objc class Items: NSObject {
    @objc static let sharedInstance = Items()
    
    @objc private(set) var allItems: [Item] = []
    
    private override init() {
        super.init()
        allItems = loadItemsFromJSON()
    }
    
    private func loadItemsFromJSON() -> [Item] {
        guard let filePath = Bundle.main.path(forResource: "data", ofType: "json"),
              let jsonData = try? Data(contentsOf: URL(fileURLWithPath: filePath)),
              let jsonArray = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [[String: Any]] else {
            return []
        }
        
        var itemArray: [Item] = []
        
        for itemDictionary in jsonArray {
            let itemObject = Item()
            
            if let id = itemDictionary["id"] as? NSNumber {
                itemObject.ID = id
            }
            if let sale = itemDictionary["Sale"] as? NSNumber {
                itemObject.sale = sale
            }
            if let productName = itemDictionary["ProductName"] as? String {
                itemObject.itemName = productName
            }
            if let url = itemDictionary["URL"] as? String {
                itemObject.photoURL = url
            }
            if let price = itemDictionary["Price"] as? Double {
                itemObject.price = price
            } else if let priceNumber = itemDictionary["Price"] as? NSNumber {
                itemObject.price = priceNumber.doubleValue
            }
            if let productType = itemDictionary["ProductType"] as? String {
                itemObject.itemCategory = productType
            }
            if let brandName = itemDictionary["BrandName"] as? String {
                itemObject.brand = brandName
            }
            if let quality = itemDictionary["Quality"] as? String {
                itemObject.quality = quality
            }
            itemObject.cartAddedQuantity = 0
            
            itemArray.append(itemObject)
        }
        
        return itemArray
    }
}
